src/math/bvh.js

```diff
@@ -0,0 +1,456 @@
+            const leftArea = Array(this.bins - 1);
+            const rightArea = Array(this.bins - 1);
+            const leftCount = Array(this.bins - 1);
+            const rightCount = Array(this.bins - 1);
+
+            const leftBox = new BoundingBox(new Vec3(), new Vec3());
+            const rightBox = new BoundingBox(new Vec3(), new Vec3());
+            let leftSum = 0;
+            let rightSum = 0;
+
+            for (let i = 0; i < this.bins - 1; i++) {
+                leftSum += bin[i].triCount;
+                leftCount[i]  = leftSum;
+                leftBox.setMinMax(vec3Min(leftBox.getMin(), bin[i].bounds.getMin()), vec3Max(leftBox.getMax(), bin[i].bounds.getMax()));
+                leftArea[i] = leftBox.area();
+                rightSum += bin[this.bins - 1 - i].triCount;
+                rightCount[this.bins - 2 - i] = rightSum;
+                rightBox.setMinMax(vec3Min(rightBox.getMin(), bin[this.bins - 1 - i].bounds.getMin()), vec3Max(rightBox.getMax(), bin[this.bins - 1 - i].bounds.getMax()));
+                rightArea[this.bins - 2 - i] = rightBox.area();
+            }
+
+            scale = (boundsMax - boundsMin) / this.bins;
+            for (let i = 0; i < this.bins - 1; i++) {
+                const planeCost = leftCount[i] * leftArea[i] + rightCount[i] * rightArea[i];
+                if (planeCost < bestCost) {
+                    splitDetails.axis = b;
+                    splitDetails.splitPos = boundsMin + scale * (i + 1);
+                    bestCost = planeCost;
+                }
+            }
+        }
+        return bestCost;
+    }
+
+    /**
+     * Calculates the estimated cost of raycasting through a node
+     *
+     * @param {BVHNode} node - The node to be evaluated
+     * @returns {number} The cost of the raycast
+     * @ignore
+     */
+    calculateNodeCost(node) {
+        const extent = new Vec3();
+        extent.sub2(node.aabbMax, node.aabbMin);
+
+        const surfaceArea = extent.x * extent.y + extent.y * extent.z + extent.z * extent.x;
+
+        return node.triCount * surfaceArea;
+    }
+
+    /**
+     * Subdivides the BVH on a node
+     *
+     * @param {number} nodeIdx - The index of the node to subdivide at
+     * @ignore
+     */
+    Subdivide(nodeIdx) {
+        const node = this.bvhNode[nodeIdx];
+
+        const { triangles, triIdx } = this;
+
+        const splitDetails = { node: node, axis: null, splitPos: null, splitCost: null };
+
+        // const splitCost = this.findBestSplitPlane(splitDetails);
+        this.findBestSplitPlane(splitDetails);
+
+        const axis = splitDetails.axis;
+
+        const splitPos = splitDetails.splitPos;
+
+        // const noSplitCost = this.calculateNodeCost(node);
+
+        // terminates tree building based on split cost
+        // if (splitCost >= noSplitCost) {
+        //     return;
+        // }
+
+        // terminates tree building when number of leafnodes is less than 5
+        if (node.triCount < 10) {
```

Comment:
5 or 10?

Reply:
Changed the comment to say threshold; should I make this configurable?

Reply:
probably not at this stage, a fixed number should be ok